webhook_app/1.0.0/functions/server/sharepoint.js

```diff
@@ -0,0 +1,266 @@
+ */
+function processRequest(data) {
+    return data.map(event => ({
+        executionContext: {userData: {subscriptionId: event.subscriptionId}},
+        body: event
+    }));
+}
+
+/**
+ * echoes validation token
+ *
+ * @param {string} token the token SP send on webhook registration
+ * @return {Promise<{data: {response: {statusCode: number, content: *}}}>} response to build HTTP response to SP
+ */
+function confirmSubscription(token) {
+    return Promise.resolve({
+        data: {
+            response: {statusCode: 200, content: token}
+        }
+    });
+}
+
+/**
+ * get changes since the token. NOT IMPLEMENTED YET!
+ *
+ * @param {Object} webhook UserData from webhook and subscription
+ * @param {Object} payload a single item from getEvents() list
+ * @param {Object} session from execution context
+ * @param {Object} role from execution context
+ * @return {Promise.<Response>} result of fetch
+ *
+ * @example of webhook
+ *  { tenant: 'devloop',
+ *    listId: '0aee2504-34aa-4e06-b7fe-ec671fa78a15',
+ *    changeQuery: { Add: 'True', Item: 'True' },
+ *    changeToken: '1;3;0aee2504-34aa-4e06-b7fe-ec671fa78a15;636528416563470000;160750288',
+ *    subscriptionId: 'bf7e7782-371c-4bb0-81d0-fc8e7ce0cb57',
+ *  }
+ */
+function getChanges(webhook, payload, session, role) {
+    // dummy response to test changeToken update
+    session.userName = "test session update " + Date.now();
+    return session.store()
+        .then(() => ({
+            changeToken: 'new ChangeToken' + Date.now(),
+            changes: payload
+        })
+    );
+    // instead we need to get latest changes from SP changelog:
+    const request = new YMSRequest({
+        protocol: "https",
+        host: webhook.tenant + ".sharepoint.com",
+        tenant: webhook.tenant
+    }, session, role);
+    const requestBody = {
+        "query": {
+            "__metadata": {
+                "type": "SP.ChangeLogItemQuery",
+                "ChangeToken": webhook.changeToken
+            }
+        }
+    };
+    // couldn't find a way to query SP change log. The best of my attempts returns XML which cause exception in YMSRequest response parser:
+    return request.POST("/_api/web/lists('" + webhook.listId + "')/GetListItemChangesSinceToken", requestBody, {"Content-Type": "application/json; odata=verbose"})
+        .then(result => {
+            return result;
+        }).catch(e => {
+            throw e;
+        });
+}
+
+exports.REST = {
+
+    /**
+     * register webhook on SP API
+     *
+     * @param {Object} _args command line parameters
+     * @param {Object} _options request options and post data
+     * @param {YServer} _serv server
+     * @param {YRole} _role from front API
+     * @param {YSession} _session from front API
+     * @return {Promise.<Object>} response and userData to return to the client
+     */
+    register(_args, _options, _serv, _role, _session) {
+        const webhook = JSON.parse(_options.data);

```

Comment:
are you sure `_options.data` will always be a parsed ok?
i've had inconsistent behaviour — sometimes it comes as string and all is good but others it's already proper JSON and fails when parsing... and from the jsdoc the functions seems to expect an object in `_options`... ;)

Reply:
Yes, I believe so. At least I couldn't find a way to query server function with a raw unstringified json object via fetch API. The whole body of the request is stringified in https://github.com/YupanaInc/Yupana-Framework/blob/b7159d7/server/lib/YFetch.js#L506. The only way to send raw json is to make a direct HTTP request to the server function like `curl ...... --data-binary "{.....}"`, which is not a legitimate way to use it.

Even tho the function is exposed, it shouldn't be called directly. The client should call [YWebHook.js/register](https://github.com/YupanaInc/Yupana-Framework/blob/5ed31e8/server/lib/YWebHook.js#L234) instead. It validates the input and [executes the server function](https://github.com/YupanaInc/Yupana-Framework/blob/5ed31e8/server/lib/YWebHook.js#L282) using fetch API, which should guarantee proper data format unless I am missing something.